test/extended/networking/network_segmentation.go

```diff
@@ -0,0 +1,862 @@
+						continue
+					}
+					// positive case for UDN pod is a successful healthcheck, checked later
+					By("checking the default network pod can't reach UDN pod on IP " + destIP)
+					Consistently(func() bool {
+						return connectToServer(podConfiguration{namespace: defaultPod.Namespace, name: defaultPod.Name}, destIP, port) != nil
+					}, 5*time.Second).Should(BeTrue())
+				}
+
+				defaultIPv4, defaultIPv6, err := podIPsForDefaultNetwork(
+					cs,
+					defaultPod.Namespace,
+					defaultPod.Name,
+				)
+				Expect(err).NotTo(HaveOccurred())
+
+				for _, destIP := range []string{defaultIPv4, defaultIPv6} {
+					if destIP == "" {
+						continue
+					}
+					By("checking the default network client pod can reach default pod on IP " + destIP)
+					Eventually(func() bool {
+						return connectToServer(podConfiguration{namespace: defaultClientPod.Namespace, name: defaultClientPod.Name}, destIP, defaultPort) == nil
+					}).Should(BeTrue())
+					By("checking the UDN pod can't reach the default network pod on IP " + destIP)
+					Consistently(func() bool {
+						return connectToServer(udnPodConfig, destIP, defaultPort) != nil
+					}, 5*time.Second).Should(BeTrue())
+				}
+
+				By("asserting healthcheck works (kubelet can access the UDN pod)")
+				// The pod should be ready
+				Expect(podutils.IsPodReady(udnPod)).To(BeTrue())
+
+				// connectivity check is run every second + 1sec initialDelay
+				// By this time we have spent at least 8 seconds doing the above checks
+				udnPod, err = cs.CoreV1().Pods(udnPod.Namespace).Get(context.Background(), udnPod.Name, metav1.GetOptions{})
+				Expect(err).NotTo(HaveOccurred())
+				Expect(udnPod.Status.ContainerStatuses[0].RestartCount).To(Equal(int32(0)))
+
+				// TODO
+				//By("checking non-kubelet default network host process can't reach the UDN pod")
+
+				By("asserting UDN pod can't reach host via default network interface")
+				// tweak pod route to use default network interface as default
+				podAnno, err := unmarshalPodAnnotation(udnPod.Annotations, "default")
+				Expect(err).NotTo(HaveOccurred())
+				for _, podIP := range podAnno.IPs {
+					ipCommand := []string{"exec", udnPod.Name, "--", "ip"}
+					if podIP.IP.To4() == nil {
+						ipCommand = append(ipCommand, "-6")
+					}
+					// 1. Find current default route and delete it
+					defRoute, err := e2ekubectl.RunKubectl(udnPod.Namespace,
+						append(ipCommand, "route", "show", "default")...)
+					Expect(err).NotTo(HaveOccurred())
+					defRoute = strings.TrimSpace(defRoute)
+					if defRoute == "" {
+						continue
+					}
+					framework.Logf("Found default route %v, deleting", defRoute)
+					cmd := append(ipCommand, "route", "del")
+					_, err = e2ekubectl.RunKubectl(udnPod.Namespace,
+						append(cmd, strings.Split(defRoute, " ")...)...)
+					Expect(err).NotTo(HaveOccurred())
+
+					// 2. Add a new default route to use default network interface
+					gatewayIP := NextIP(Network(podIP).IP)
```

Comment:
to avoid vendoring a LOT of changes I've copied this code directly from `containernetworking/plugins`, which has an apache 2.0 license.

Still, is there any reason for us to require using this function ? Is there another alternative - e.g. using an hard-coded IP, etc ? @npinaeva (tagging you since you've wrote this test upstream).

Reply:
After syncing offline, we reached the conclusion that we **could** ensure one of the default network pods are scheduled in the same node as the UDN pod, parse its annotations, and grab its gateway IP (we would need that for layer3, where each node has a different subnet).

Seems too much work - let's keep this copied code for now.